src/inventorymanager.cpp

```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
 {
 	ServerScripting *sa = PLAYER_TO_SA(player);
-	int src_can_take_count = 0xffff;
-	if (from_inv.type == InventoryLocation::DETACHED)
-		src_can_take_count = sa->detached_inventory_AllowMove(*this, try_take_count, player);
-	else if (from_inv.type == InventoryLocation::NODEMETA)
-		src_can_take_count = sa->nodemeta_inventory_AllowMove(*this, try_take_count, player);
-	else if (from_inv.type == InventoryLocation::PLAYER)
-		src_can_take_count = sa->player_inventory_AllowMove(*this, try_take_count, player);
+	int can_take_count = 0xffff;
+	if (ma.from_inv.type == InventoryLocation::DETACHED)
+		can_take_count = sa->detached_inventory_AllowMove(ma, try_take_count, player);
+	else if (ma.from_inv.type == InventoryLocation::NODEMETA)
+		can_take_count = sa->nodemeta_inventory_AllowMove(ma, try_take_count, player);
+	else if (ma.from_inv.type == InventoryLocation::PLAYER)
+		can_take_count = sa->player_inventory_AllowMove(ma, try_take_count, player);
 	else
 		assert(false);
-	return src_can_take_count;
+	return can_take_count;
+}
+
+static std::pair<s16, bool> move(const MoveAction& ma, InventoryList *list_from, InventoryList *list_to,
+	u16 count, bool ignore_rollback, InventoryManager *mgr, ServerActiveObject *player, IGameDef *gamedef)
+{
+	ItemStack src_item = list_from->getItem(ma.from_i);
+
+	if (src_item.empty())
+		return {0, false};
+
+	if (count == 0 || src_item.count < count)
+		count = src_item.count;
+
+	bool inf_from = false;
+	bool inf_to = false;
+	u16 orig_count = count;
+
+	/*
+		Collect information of endpoints
+	*/
+	if (ma.from_inv == ma.to_inv) {
+		// Move action within the same inventory
+		int can_take_count = allowMove(ma, count, player);
+		if (can_take_count >= 0 && can_take_count < count)
+			count = can_take_count;
+	} else {
+		src_item.count = count;
+		int can_take_count = allowTake(ma, src_item, player);
+		if (can_take_count == -1)
+			inf_from = true;
+		else if (can_take_count >= 0 && can_take_count < count)
+			count = can_take_count;
+
+		if (count > 0) {
+			src_item.count = count;
+			int can_put_count = allowPut(ma, src_item, player);
+			if (can_put_count == -1)
+				inf_to = true;
+			else if (can_put_count >= 0 && can_put_count < count)
+				count = can_put_count;
+		}
+	}
+
+	u16 moved_count = 0;
+	if (count > 0)
+		moved_count = list_from->moveItem(ma.from_i, list_to, ma.to_i, count);
+
+	infostream << "IMoveAction::apply(): moved " << moved_count << " from " << count << " items:"
+		<< " orig_count=" << orig_count
+		<< " from inv=\"" << ma.from_inv.dump() << "\""
+		<< " list=\"" <<  ma.from_list << "\""
+		<< " i=" <<  ma.from_i
+		<< " to inv=\"" <<  ma.to_inv.dump() << "\""
+		<< " list=\"" <<  ma.to_list << "\""
+		<< " i=" << ma.to_i
+		<< std::endl;
+
+	if (moved_count == 0)
+		return {0, count == orig_count}; // Can swap only full stack
+
+	ItemStack item = list_to->getItem(ma.to_i);
+	item.count = moved_count;
+
+	if (inf_from) {
+		ItemStack& item = list_from->getItem(ma.from_i);
```

Comment:
please don't shadow `item`, that's confusing

Reply:
Fixed